### PR TITLE
[feat](job)Implementing Job in Nereids

### PR DESCRIPTION
**Pull request by CalvinKirs**
## Proposed changes

The JOB's execution SQL is currently defined by an older CUP file, which causes some issues with lexical analysis in the new optimizer as it doesn't pass under the old optimizer. Since the JOB's underlying execution already uses the new optimizer, we're planning to fully migrate to ANTLR4 for consistency.

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/info/CreateJobInfo.java [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+        // If the comment is not empty, extract the SQL statement from the end position of the comment.
+        if (org.apache.commons.lang3.StringUtils.isNotBlank(comment)) {
+
+            String lowerCaseComment = comment.toLowerCase();
+            int splitDoIndex = subSqlStmt.indexOf(lowerCaseComment) + lowerCaseComment.length();
+            subSqlStmt = subSqlStmt.substring(splitDoIndex);
+            originSubSqlStmt = originSubSqlStmt.substring(splitDoIndex);
+        }
+        // Find the position of the "do" keyword and extract the execution SQL statement from this position.
+        int executeSqlIndex = subSqlStmt.indexOf("do");
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/info/CreateJobInfo.java:223]: I think when using antler, we can get the sql string directly?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/info/CreateJobInfo.java [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+     */
+    private static String parseExecuteSql(String sql, String jobName, String comment) throws AnalysisException {
+        String lowerCaseSql = sql.toLowerCase();
+        String lowerCaseJobName = jobName.toLowerCase();
+        // Find the end position of the job name in the SQL statement.
+        int jobNameEndIndex = lowerCaseSql.indexOf(lowerCaseJobName) + lowerCaseJobName.length();
+        String subSqlStmt = lowerCaseSql.substring(jobNameEndIndex);
+        String originSubSqlStmt = sql.substring(jobNameEndIndex);
+        // If the comment is not empty, extract the SQL statement from the end position of the comment.
+        if (org.apache.commons.lang3.StringUtils.isNotBlank(comment)) {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/info/CreateJobInfo.java:215]: use `Strings.isNullOrEmpty()` ? in Guava

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/insert/InsertIntoTableCommand.java [hunk 3]
```diff
@@ -123,12 +123,17 @@ public void runWithUpdateInfo(ConnectContext ctx, StmtExecutor executor,
...
 
+    public AbstractInsertExecutor initPlan(ConnectContext ctx, StmtExecutor executor) throws Exception {
+        return initPlan(ctx, executor, true);
+    }
+
     /**
      * This function is used to generate the plan for Nereids.
      * There are some load functions that only need to the plan, such as stream_load.
      * Therefore, this section will be presented separately.
      */
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/insert/InsertIntoTableCommand.java:134]: Add comment for `needBeginTransaction`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/info/CreateJobInfo.java [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+                        comment,
+                        ConnectContext.get().getCurrentUserIdentity(),
+                        jobExecutionConfiguration,
+                        System.currentTimeMillis(),
+                        sql);
+            } catch (Exception e) {
+                throw new AnalysisException(e.getMessage());
+            }
+        } else {
+            throw new AnalysisException("Not support this sql : " + sql);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/info/CreateJobInfo.java:261]: I think we can also print the class name of this sql too

File: fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4 [hunk 5]
```diff
@@ -102,7 +103,17 @@ materializedViewStatement
...
+jobScheduleStatement
+    : CREATE JOB label=multipartIdentifier ON SCHEDULE
+        (
+            (EVERY timeInterval=INTEGER_VALUE timeUnit=identifier
+            (STARTS (startTime=STRING_LITERAL | CURRENT_TIMESTAMP))?
+            (ENDS endsTime=STRING_LITERAL)?)
+            |
+            (AT (atTime=STRING_LITERAL | CURRENT_TIMESTAMP)))
+        commentSpec?
+        DO insertIntoTableStatement                                                                #createScheduledJob
```
[the remark this answers is not in the document]
Comment (CalvinKirs) [reply fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4]: yup~

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/info/CreateJobInfo.java (hunk 1 above)
Comment (CalvinKirs) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/info/CreateJobInfo.java]: I tried it, and it works, it's awesome

**Review by morningman — DISMISSED** [review]
LGTM

**Comment by CalvinKirs** [comment]
run buildall

**Comment by CalvinKirs** [comment]
run buildall

**Comment by CalvinKirs** [comment]
run buildall

**Comment by CalvinKirs** [comment]
run buildall

**Review by 924060929 — DISMISSED** [review]

**Comment by CalvinKirs** [comment]
run buildall

File: regression-test/suites/job_p0/test_base_insert_job.groovy [hunk 6]
```diff
@@ -26,6 +26,9 @@ suite("test_base_insert_job") {
     def tableName = "t_test_BASE_inSert_job"
     def jobName = "insert_recovery_test_base_insert_job"
     def jobMixedName = "Insert_recovery_Test_base_insert_job"
+    sql """
+      SET enable_fallback_to_original_planner=false;
+    """
```
Comment (morrySnow) [inline regression-test/suites/job_p0/test_base_insert_job.groovy:31]: this variable has been removed and u do not need to add it

**Comment by CalvinKirs** [comment]
run performance

File: fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4 [hunk 7]
```diff
@@ -102,7 +103,17 @@ materializedViewStatement
     | CANCEL MATERIALIZED VIEW TASK taskId=INTEGER_VALUE ON mvName=multipartIdentifier          #cancelMTMVTask
     | SHOW CREATE MATERIALIZED VIEW mvName=multipartIdentifier                                  #showCreateMTMV
     ;
-
+jobScheduleStatement
+    : CREATE JOB label=multipartIdentifier ON SCHEDULE
```
Comment (morrySnow) [inline fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4:107]: remove create job syntax in `sql_parser.cup` and remove `CreateJobStmt.java`

Comment (CalvinKirs) [reply fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4:107]: I want to directly pick to 2.1, as 2.1 allows for a rollback. Therefore, this change will be provided in the next PR.:)

**Review by morningman — APPROVED** [review]
LGTM

**Comment by morningman** [comment]
run buildall

**Comment by CalvinKirs** [comment]
run feut

**Review by morningman — APPROVED** [review]
LGTM

**Comment by CalvinKirs** [comment]
run cloud_p0

**Review by starocean999 — APPROVED** [review]

**Pull request merged**